File: packages/decision_engine/semantic_render.py
"""Deterministic SemanticExplanation renderer (Phase 3F1).

Copy infrastructure only. No Calendar / Ask / Compare / Find UI wiring.
No LLM. No network. Does not change scores, class, ranking, or commands.
"""

from __future__ import annotations

import re
from typing import Any, Literal, Mapping

from pydantic import BaseModel, ConfigDict, Field

from packages.decision_engine.i18n.catalog import (
    SUPPORTED_LOCALES,
    required_slots,
    template_for,
    text_direction,
)
from packages.decision_engine.semantic_explanation import SemanticExplanation

RENDER_SCHEMA_VERSION = "semantic_render.v1-shadow"
SEMANTIC_STATUS = "experimental_shadow"

UNAVAILABLE_UNSUPPORTED_LOCALE = "render.unsupported_locale"
UNAVAILABLE_MISSING_ENTRY = "render.missing_catalog_entry"
UNAVAILABLE_MISSING_ARGS = "render.missing_placeholder_args"
UNAVAILABLE_MALFORMED_TEMPLATE = "render.malformed_template"

_SLOT_RE = re.compile(r"\{([a-z][a-z0-9_]*)\}")
_UNSAFE_BRACE_RE = re.compile(r"[{}]")


class SemanticRenderError(ValueError):
    """Deterministic renderer failure (strict mode)."""

    def __init__(self, code: str, detail: str) -> None:
        super().__init__(f"{code}: {detail}")
        self.code = code
        self.detail = detail


class SemanticRenderContext(BaseModel):
    """Optional display labels. Never inferred from ranking."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    option_labels: dict[str, str] = Field(default_factory=dict)
    left_label: str | None = None
    right_label: str | None = None


class RenderedSemanticExplanation(BaseModel):
    """Localized copy slots. Not a command. Not a ranking."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    schema_version: Literal["semantic_render.v1-shadow"] = RENDER_SCHEMA_VERSION
    semantic_status: Literal["experimental_shadow"] = SEMANTIC_STATUS
    locale: str
    text_direction: Literal["ltr", "rtl"]
    status: Literal["ok", "unavailable"]
    unavailable_code: str | None = None
    headline: str | None = None
    summary: str | None = None
    opportunity: str | None = None
    posture: str | None = None
    tradeoff: str | None = None
    supports: tuple[str, ...] = ()
    cautions: tuple[str, ...] = ()
    safety: tuple[str, ...] = ()


def _sanitize_slot_value(value: Any) -> str:
    text = str(value).strip()
    return _UNSAFE_BRACE_RE.sub("", text)


def _as_explanation(
    explanation: SemanticExplanation | Mapping[str, Any],
) -> SemanticExplanation:
    if isinstance(explanation, SemanticExplanation):
        return explanation
    return SemanticExplanation.model_validate(dict(explanation))


def _as_context(
    display_context: SemanticRenderContext | Mapping[str, Any] | None,
) -> SemanticRenderContext:
    if display_context is None:
        return SemanticRenderContext()
    if isinstance(display_context, SemanticRenderContext):
        return display_context
    return SemanticRenderContext.model_validate(dict(display_context))


def _label_for(
    option_id: str | None,
    labels: Mapping[str, str],
) -> str | None:
    if not option_id:
        return None
    if option_id in labels and labels[option_id]:
        return _sanitize_slot_value(labels[option_id])
    return _sanitize_slot_value(option_id)


def build_named_slots(
    explanation: SemanticExplanation,
    display_context: SemanticRenderContext | None = None,
) -> dict[str, str]:
    """Named interpolation slots. Meaning is never positional."""
    ctx = display_context or SemanticRenderContext()
    args = explanation.localization_args or {}
    labels = dict(ctx.option_labels)
    left_id = args.get("left_id")
    right_id = args.get("right_id")
    if ctx.left_label and left_id:
        labels.setdefault(str(left_id), ctx.left_label)
    if ctx.right_label and right_id:
        labels.setdefault(str(right_id), ctx.right_label)

    left_id_s = str(left_id) if left_id else None
    right_id_s = str(right_id) if right_id else None
    left_label = _label_for(left_id_s, labels)
    right_label = _label_for(right_id_s, labels)

    def resolve(option_id: Any) -> str | None:
        if option_id is None or option_id == "":
            return None
        key = str(option_id)
        if key == left_id_s:
            return left_label
        if key == right_id_s:
            return right_label
        return _label_for(key, labels)

    slots: dict[str, str] = {}
    if left_label:
        slots["left_label"] = left_label
    if right_label:
        slots["right_label"] = right_label
    higher = resolve(args.get("score_preference"))
    cleaner = resolve(args.get("posture_preference"))
    if higher:
        slots["higher_score_option"] = higher
    if cleaner:
        slots["cleaner_posture_option"] = cleaner
        slots["option"] = cleaner
    return slots


def interpolate(template: str, slots: Mapping[str, str]) -> str:
    missing: list[str] = []

    def replace(match: re.Match[str]) -> str:
        name = match.group(1)
        if name not in slots or slots[name] == "":
            missing.append(name)
            return match.group(0)
        return slots[name]

    rendered = _SLOT_RE.sub(replace, template)
    if missing:
        raise SemanticRenderError(
            UNAVAILABLE_MISSING_ARGS,
            "missing named slots: " + ", ".join(sorted(set(missing))),
        )
    leftover = _SLOT_RE.findall(rendered)
    if leftover or "{" in rendered or "}" in rendered:
        raise SemanticRenderError(
            UNAVAILABLE_MALFORMED_TEMPLATE,
            f"unresolved placeholders in {template!r}",
        )
    return rendered


def _render_code(code: str | None, locale: str, slots: Mapping[str, str]) -> str | None:
    if not code:
        return None
    needed = required_slots(code)
    missing = [name for name in needed if not slots.get(name)]
    if missing:
        raise SemanticRenderError(
            UNAVAILABLE_MISSING_ARGS,
            f"{code} missing {missing}",
        )
    try:
        template = template_for(locale, code)
    except KeyError as exc:
        raise SemanticRenderError(UNAVAILABLE_MISSING_ENTRY, str(exc)) from exc
    return interpolate(template, slots)


def _unavailable(
    locale: str,
    code: str,
    *,
    strict: bool,
    detail: str,
) -> RenderedSemanticExplanation:
    if strict:
        raise SemanticRenderError(code, detail)
    direction: Literal["ltr", "rtl"]
    direction = "rtl" if text_direction(locale) == "rtl" else "ltr"
    return RenderedSemanticExplanation(
        locale=locale,
        text_direction=direction,
        status="unavailable",
        unavailable_code=code,
    )


def render_semantic_explanation(
    explanation: SemanticExplanation | Mapping[str, Any],
    locale: str,
    display_context: SemanticRenderContext | Mapping[str, Any] | None = None,
    *,
    strict: bool = True,
) -> RenderedSemanticExplanation:
    """Render codes to localized copy. Pure. Does not mutate explanation."""
    if locale not in SUPPORTED_LOCALES:
        return _unavailable(
            locale,
            UNAVAILABLE_UNSUPPORTED_LOCALE,
            strict=strict,
            detail=f"unsupported locale {locale!r}",
        )
    try:
        model = _as_explanation(explanation)
        context = _as_context(display_context)
        slots = build_named_slots(model, context)
        headline = _render_code(model.headline_code, locale, slots)
        summary = _render_code(model.summary_code, locale, slots)
        opportunity = _render_code(model.opportunity_code, locale, slots)
        posture = _render_code(model.posture_code, locale, slots)
        tradeoff = _render_code(model.tradeoff_code, locale, slots)
        supports = tuple(
            _render_code(code, locale, slots) or "" for code in model.support_codes
        )
        cautions = tuple(
            _render_code(code, locale, slots) or "" for code in model.caution_codes
        )
        safety = tuple(
            _render_code(code, locale, slots) or "" for code in model.safety_codes
        )
    except SemanticRenderError as exc:
        return _unavailable(locale, exc.code, strict=strict, detail=exc.detail)

    direction: Literal["ltr", "rtl"]
    direction = "rtl" if text_direction(locale) == "rtl" else "ltr"
    return RenderedSemanticExplanation(
        locale=locale,
        text_direction=direction,
        status="ok",
        headline=headline,
        summary=summary,
        opportunity=opportunity,
        posture=posture,
        tradeoff=tradeoff,
        supports=supports,
        cautions=cautions,
        safety=safety,
    )


__all__ = [
    "RENDER_SCHEMA_VERSION",
    "RenderedSemanticExplanation",
    "SEMANTIC_STATUS",
    "SemanticRenderContext",
    "SemanticRenderError",
    "build_named_slots",
    "interpolate",
    "render_semantic_explanation",
]

File: apps/web/lib/decision-intelligence/semantic-render.ts
/**
 * Deterministic SemanticExplanation renderer (Phase 3F1 catalogs).
 * Same JSON catalogs as packages/decision_engine/i18n/catalogs.
 * No LLM. No English fallback for a supported locale.
 */

import arCatalog from './catalogs/ar.json';
import enCatalog from './catalogs/en.json';
import faCatalog from './catalogs/fa.json';
import ruCatalog from './catalogs/ru.json';
import type {
  RenderedSemanticExplanation,
  SemanticExplanationInput,
  SemanticPreviewLocale,
  SemanticRenderContext,
} from './types';

export const SUPPORTED_LOCALES = ['en', 'fa', 'ar', 'ru'] as const;

export const REQUIRED_SLOTS: Record<string, readonly string[]> = {
  'semantic.higher_score_stronger_opportunity': ['higher_score_option'],
  'semantic.cleaner_posture': ['option'],
  'semantic.lower_score_cleaner_posture': ['cleaner_posture_option'],
  'semantic.material_tradeoff': [
    'higher_score_option',
    'cleaner_posture_option',
  ],
  'semantic.near_tie_cleaner_posture': ['option'],
  'semantic.score_advantage_with_caution': ['higher_score_option'],
};

const CATALOGS = {
  en: enCatalog,
  fa: faCatalog,
  ar: arCatalog,
  ru: ruCatalog,
} as const;

const TEXT_DIRECTION: Record<string, 'ltr' | 'rtl'> = {
  en: 'ltr',
  fa: 'rtl',
  ar: 'rtl',
  ru: 'ltr',
};

export const UNAVAILABLE_UNSUPPORTED_LOCALE = 'render.unsupported_locale';
export const UNAVAILABLE_MISSING_ENTRY = 'render.missing_catalog_entry';
export const UNAVAILABLE_MISSING_ARGS = 'render.missing_placeholder_args';
export const UNAVAILABLE_MALFORMED_TEMPLATE = 'render.malformed_template';

const SLOT_RE = /\{([a-z][a-z0-9_]*)\}/g;
const UNSAFE_BRACE_RE = /[{}]/g;

type CatalogFile = {
  locale: string;
  dir: string;
  messages: Record<string, string>;
  posture_terms: Record<string, string>;
};

function catalogFor(locale: string): CatalogFile | null {
  if (!(locale in CATALOGS)) return null;
  return CATALOGS[locale as SemanticPreviewLocale] as CatalogFile;
}

function sanitize(value: unknown): string {
  return String(value ?? '')
    .trim()
    .replace(UNSAFE_BRACE_RE, '');
}

function requiredSlots(code: string): readonly string[] {
  return REQUIRED_SLOTS[code] ?? [];
}

export function textDirection(locale: string): 'ltr' | 'rtl' {
  return TEXT_DIRECTION[locale] ?? 'ltr';
}

export function templateFor(locale: string, code: string): string | null {
  const catalog = catalogFor(locale);
  const template = catalog?.messages[code];
  if (!template || !template.trim()) return null;
  return template;
}

function interpolate(
  template: string,
  slots: Record<string, string>
): { ok: true; text: string } | { ok: false; code: string } {
  const missing: string[] = [];
  const rendered = template.replace(SLOT_RE, (_all, name: string) => {
    if (!slots[name]) {
      missing.push(name);
      return `{${name}}`;
    }
    return slots[name];
  });
  if (missing.length) {
    return { ok: false, code: UNAVAILABLE_MISSING_ARGS };
  }
  if (SLOT_RE.test(rendered) || rendered.includes('{') || rendered.includes('}')) {
    SLOT_RE.lastIndex = 0;
    return { ok: false, code: UNAVAILABLE_MALFORMED_TEMPLATE };
  }
  SLOT_RE.lastIndex = 0;
  return { ok: true, text: rendered };
}

export function buildNamedSlots(
  explanation: SemanticExplanationInput,
  displayContext: SemanticRenderContext = {}
): Record<string, string> {
  const args = explanation.localization_args ?? {};
  const labels = { ...(displayContext.option_labels ?? {}) };
  const leftId = args.left_id != null ? String(args.left_id) : '';
  const rightId = args.right_id != null ? String(args.right_id) : '';
  if (displayContext.left_label && leftId) {
    labels[leftId] ??= displayContext.left_label;
  }
  if (displayContext.right_label && rightId) {
    labels[rightId] ??= displayContext.right_label;
  }

  const labelFor = (optionId: unknown): string | null => {
    if (optionId == null || optionId === '') return null;
    const key = String(optionId);
    if (labels[key]) return sanitize(labels[key]);
    return sanitize(key);
  };

  const slots: Record<string, string> = {};
  const leftLabel = labelFor(leftId || null);
  const rightLabel = labelFor(rightId || null);
  if (leftLabel) slots.left_label = leftLabel;
  if (rightLabel) slots.right_label = rightLabel;
  const higher = labelFor(args.score_preference);
  const cleaner = labelFor(args.posture_preference);
  if (higher) slots.higher_score_option = higher;
  if (cleaner) {
    slots.cleaner_posture_option = cleaner;
    slots.option = cleaner;
  }
  return slots;
}

function renderCode(
  code: string | null | undefined,
  locale: string,
  slots: Record<string, string>
): { ok: true; text: string | null } | { ok: false; code: string } {
  if (!code) return { ok: true, text: null };
  for (const name of requiredSlots(code)) {
    if (!slots[name]) {
      return { ok: false, code: UNAVAILABLE_MISSING_ARGS };
    }
  }
  const template = templateFor(locale, code);
  if (!template) return { ok: false, code: UNAVAILABLE_MISSING_ENTRY };
  const interpolated = interpolate(template, slots);
  if (!interpolated.ok) return interpolated;
  return { ok: true, text: interpolated.text };
}

function unavailable(
  locale: string,
  code: string
): RenderedSemanticExplanation {
  return {
    schema_version: 'semantic_render.v1-shadow',
    semantic_status: 'experimental_shadow',
    locale,
    text_direction: textDirection(locale),
    status: 'unavailable',
    unavailable_code: code,
    headline: null,
    summary: null,
    opportunity: null,
    posture: null,
    tradeoff: null,
    supports: [],
    cautions: [],
    safety: [],
  };
}

export function renderSemanticExplanation(
  explanation: SemanticExplanationInput | null | undefined,
  locale: string,
  displayContext: SemanticRenderContext = {}
): RenderedSemanticExplanation {
  if (!explanation) {
    return unavailable(locale, 'render.missing_explanation');
  }
  if (!SUPPORTED_LOCALES.includes(locale as SemanticPreviewLocale)) {
    return unavailable(locale, UNAVAILABLE_UNSUPPORTED_LOCALE);
  }
  const slots = buildNamedSlots(explanation, displayContext);
  const fields = {
    headline: renderCode(explanation.headline_code, locale, slots),
    summary: renderCode(explanation.summary_code, locale, slots),
    opportunity: renderCode(explanation.opportunity_code, locale, slots),
    posture: renderCode(explanation.posture_code, locale, slots),
    tradeoff: renderCode(explanation.tradeoff_code, locale, slots),
  };
  for (const result of Object.values(fields)) {
    if (!result.ok) return unavailable(locale, result.code);
  }
  const supports: string[] = [];
  const cautions: string[] = [];
  const safety: string[] = [];
  for (const code of explanation.support_codes ?? []) {
    const result = renderCode(code, locale, slots);
    if (!result.ok) return unavailable(locale, result.code);
    if (result.text) supports.push(result.text);
  }
  for (const code of explanation.caution_codes ?? []) {
    const result = renderCode(code, locale, slots);
    if (!result.ok) return unavailable(locale, result.code);
    if (result.text) cautions.push(result.text);
  }
  for (const code of explanation.safety_codes ?? []) {
    const result = renderCode(code, locale, slots);
    if (!result.ok) return unavailable(locale, result.code);
    if (result.text) safety.push(result.text);
  }
  return {
    schema_version: 'semantic_render.v1-shadow',
    semantic_status: 'experimental_shadow',
    locale,
    text_direction: textDirection(locale),
    status: 'ok',
    unavailable_code: null,
    headline: fields.headline.ok ? fields.headline.text : null,
    summary: fields.summary.ok ? fields.summary.text : null,
    opportunity: fields.opportunity.ok ? fields.opportunity.text : null,
    posture: fields.posture.ok ? fields.posture.text : null,
    tradeoff: fields.tradeoff.ok ? fields.tradeoff.text : null,
    supports,
    cautions,
    safety,
  };
}

export function catalogMessages(locale: SemanticPreviewLocale): Record<string, string> {
  return { ...(catalogFor(locale)?.messages ?? {}) };
}
